### PR TITLE
A new search API for API keys - core search function

### PR DESCRIPTION
**Pull request by ywangd**
This PR adds a new API for searching API keys. The API supports searching API keys with a controlled list of field names and a subset of Query DSL. It also provides a translation layer between the field names used in the REST layer and those in the index layer. This is to prevent tight coupling between the user facing request and index mappings so that they can evolve separately.

Compared to the Get API key API, this new search API automatically applies calling user's security context similar to regular searches, e.g. if the user has only `manage_own_api_key` privilege, only keys owned by the user are returned in the search response.

Relates: #71023
Supercedes: #73705

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ManageOwnApiKeyClusterPrivilege.java [hunk 1]
```diff
@@ -75,6 +76,10 @@ protected boolean extendedCheck(String action, TransportRequest request, Authent
                         invalidateApiKeyRequest.getUserName(), invalidateApiKeyRequest.getRealmName(),
                         invalidateApiKeyRequest.ownedByAuthenticatedUser()));
                 }
+            } else if (request instanceof QueryApiKeyRequest) {
+                final QueryApiKeyRequest queryApiKeyRequest = (QueryApiKeyRequest) request;
+                queryApiKeyRequest.setFilterForCurrentUser();
+                return true;
```
Comment (ywangd) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ManageOwnApiKeyClusterPrivilege.java:82]: As discussed, this is the simple way of applying calling user's security context. It works but is probably not a long term solution. I am open to alternatives.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java [hunk 2]
```diff
@@ -1123,6 +1125,27 @@ public void getApiKeys(String realmName, String username, String apiKeyName, Str
...
+
+        final SecurityIndexManager frozenSecurityIndex = securityIndex.freeze();
+        if (frozenSecurityIndex.indexExists() == false) {
+            wrappedListener.onResponse(Collections.emptyList());
+        } else if (frozenSecurityIndex.isAvailable() == false) {
+            wrappedListener.onFailure(frozenSecurityIndex.getUnavailableReason());
+        } else {
+            findApiKeys(apiKeyBoolQueryBuilder, true, true, wrappedListener);
+        }
+    }
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:1147]: For the time being, this uses the existing `findApiKeys` method which in turns uses scroll. It will be changed to use normal search once the API supports pagination parameters like `size`, `from`, `to` etc.

**Comment by ywangd** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/apikey/QueryApiKeyResponse.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            return (args[0] == null) ? QueryApiKeyResponse.emptyResponse() : new QueryApiKeyResponse((List<ApiKey>) args[0]);
+        });
+
+    static {
+        PARSER.declareObjectArray(optionalConstructorArg(), (p, c) -> ApiKey.fromXContent(p), new ParseField("api_keys"));
+    }
+
+    public static QueryApiKeyResponse fromXContent(XContentParser parser) throws IOException {
+        return PARSER.parse(parser, null);
+    }
```
Comment (tvernum) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/apikey/QueryApiKeyResponse.java:80]: Where is this used?

File: x-pack/plugin/security/qa/security-basic/src/javaRestTest/resources/roles.yml [hunk 4]
```diff
@@ -6,3 +6,14 @@ security_test_role:
   indices:
     - names: [ "index_allowed" ]
       privileges: [ "read", "write", "create_index" ]
+
+# The admin role also has the manage_own_api_key privilege to ensure this lesser privilege will not
+# be triggered when there is the greater manage_api_key privilege
```
Comment (tvernum) [inline x-pack/plugin/security/qa/security-basic/src/javaRestTest/resources/roles.yml:11]:
```suggestion
# interfere with the behaviour of the greater manage_api_key privilege
```

File: x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryApiKeyIT.java [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+        // An empty request body means search for all keys
+        assertQuery(API_KEY_ADMIN_AUTH_HEADER,
+            randomBoolean() ? "" : "{\"query\":{\"match_all\":{}}}",
+            apiKeys -> assertThat(apiKeys.size(), equalTo(6)));
+
+        assertQuery(API_KEY_ADMIN_AUTH_HEADER,
+            "{\"query\":{\"bool\":{\"must\":[" +
+                "{\"prefix\":{\"metadata.application\":\"fleet\"}},{\"term\":{\"metadata.environment.os\":\"Cat\"}}]}}}",
+            apiKeys -> {
+                assertThat(apiKeys.size(), equalTo(2));
```
Comment (tvernum) [inline x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryApiKeyIT.java:59]:
```suggestion
                assertThat(apiKeys, hasSize(2));
```
This produces a better failure message

File: x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryApiKeyIT.java [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+                assertThat(apiKeys.get(0).get("name"), oneOf("my-org/ingest-key-1", "my-org/management-key-1"));
+                assertThat(apiKeys.get(1).get("name"), oneOf("my-org/ingest-key-1", "my-org/management-key-1"));
```
Comment (tvernum) [inline x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryApiKeyIT.java:61]: Don't apply this directly, it probably doesn't compile:
```suggestion
                assertThat(apiKeys.stream().map(k -> k.get("name")).collect(Collectors.toList()), containsInAnyOrder("my-org/ingest-key-1", "my-org/management-key-1"));
```

File: x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryApiKeyIT.java [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+                assertThat(apiKeys.size(), equalTo(6));
+                // search using explicit IDs
+                try {
+
+                    assertQuery(API_KEY_ADMIN_AUTH_HEADER,
+                        "{ \"query\": { \"ids\": { \"values\": ["
+                            + apiKeys.stream().map(m -> "\"" + m.get("id") + "\"").collect(Collectors.joining(",")) + "] } } }",
+                        keys -> {
+                            assertThat(keys.size(), equalTo(6));
+                        });
```
Comment (tvernum) [inline x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryApiKeyIT.java:99]: Should this be a subset?
```suggestion
                    var subset = randomSubsetOf(randomIntBetween(1,5), apiKeys); 
                    assertQuery(API_KEY_ADMIN_AUTH_HEADER,
                        "{ \"query\": { \"ids\": { \"values\": ["
                            + subset.stream().map(m -> "\"" + m.get("id") + "\"").collect(Collectors.joining(",")) + "] } } }",
                        keys -> {
                            assertThat(keys, hasSize(subset.size()));
                        });
```

File: x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryApiKeyIT.java [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+        assertQueryError(API_KEY_ADMIN_AUTH_HEADER, 400,
+            "{ \"query\": { \"prefix\": {\"api_key_hash\": \"{PBKDF2}10000$\"} } }");
+
+        // Search for api keys won't return other entities
+        assertQuery(API_KEY_ADMIN_AUTH_HEADER,
+            "{ \"query\": { \"term\": {\"name\": \"someone\"} } }",
+            apiKeys -> {
+                assertThat(apiKeys, empty());
+            });
+
```
Comment (tvernum) [inline x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryApiKeyIT.java:115]: Can we also have a test for searching on fields that aren't allowed in the API but are used by the service itself (doc_type, api_key_invalidated)?

File: x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryApiKeyIT.java [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+        // User with manage_own_api_key will only see its own keys
+        assertQuery(API_KEY_USER_AUTH_HEADER,
+            randomBoolean() ? "" : "{\"query\":{\"match_all\":{}}}",
+            apiKeys -> {
+            assertThat(apiKeys.size(), equalTo(2));
+            assertThat(apiKeys.stream().map(m -> m.get("name")).collect(Collectors.toSet()),
+                equalTo(Set.of("my-ingest-key-1", "my-alert-key-2")));
```
Comment (tvernum) [inline x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryApiKeyIT.java:122]:
```suggestion
                containsInAnyOrder("my-ingest-key-1", "my-alert-key-2"));
```

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+                finalQuery.must(processedQuery);
+            } else {
+                final BoolQueryBuilder boolQueryBuilder = (BoolQueryBuilder) processedQuery;
+                finalQuery.minimumShouldMatch(boolQueryBuilder.minimumShouldMatch());
+                finalQuery.adjustPureNegative(boolQueryBuilder.adjustPureNegative());
+                boolQueryBuilder.must().forEach(finalQuery::must);
+                boolQueryBuilder.should().forEach(finalQuery::should);
+                boolQueryBuilder.mustNot().forEach(finalQuery::mustNot);
+                boolQueryBuilder.filter().forEach(finalQuery::filter);
+            }
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java:53]: Is this special case necessary? Isn't it simpler to always wrap it in a new bool so we know everything works the same way?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+            final BoolQueryBuilder newQuery =
+                QueryBuilders.boolQuery().minimumShouldMatch(query.minimumShouldMatch()).adjustPureNegative(query.adjustPureNegative());
+            query.must().stream().map(ApiKeyBoolQueryBuilder::doProcess).forEach(newQuery::must);
+            query.should().stream().map(ApiKeyBoolQueryBuilder::doProcess).forEach(newQuery::should);
+            query.mustNot().stream().map(ApiKeyBoolQueryBuilder::doProcess).forEach(newQuery::mustNot);
+            query.filter().stream().map(ApiKeyBoolQueryBuilder::doProcess).forEach(newQuery::filter);
+            return newQuery;
+        } else if (qb instanceof MatchAllQueryBuilder) {
+            return qb;
+        } else if (qb instanceof IdsQueryBuilder) {
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java:77]: Do we want to support `ids`? Isn't that exposing that API key docs have their doc id set to the api-key-id?

Maybe we need to chat about this one. I think getting the right behaviour for the API Key Id search is important.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+            return qb;
+        } else if (qb instanceof TermQueryBuilder) {
+            final TermQueryBuilder query = (TermQueryBuilder) qb;
+            final String translatedFieldName = FieldNameTranslators.translate(query.fieldName());
+            return QueryBuilders.termQuery(translatedFieldName, query.value()).caseInsensitive(query.caseInsensitive());
+        } else if (qb instanceof TermsQueryBuilder) {
+            final TermsQueryBuilder query = (TermsQueryBuilder) qb;
+            if (query.termsLookup() != null) {
+                throw new IllegalArgumentException("terms query with terms lookup is not supported for search");
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java:86]:
```suggestion
                throw new IllegalArgumentException("terms query with terms lookup is not supported for API key query");
```

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+            final WildcardQueryBuilder query = (WildcardQueryBuilder) qb;
+            final String translatedFieldName = FieldNameTranslators.translate(query.fieldName());
+            return QueryBuilders.wildcardQuery(translatedFieldName, query.value())
+                .caseInsensitive(query.caseInsensitive())
+                .rewrite(query.rewrite());
+        } else if (qb instanceof RangeQueryBuilder) {
+            final RangeQueryBuilder query = (RangeQueryBuilder) qb;
+            final String translatedFieldName = FieldNameTranslators.translate(query.fieldName());
+            if (query.relation() != null) {
+                throw new IllegalArgumentException("range query with relation is not supported for search");
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java:104]:
```suggestion
                throw new IllegalArgumentException("range query with relation is not supported for API Key query");
```

File: x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryApiKeyIT.java (hunk 8 above)
Comment (ywangd) [reply x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryApiKeyIT.java:119]: Thanks for reminding me. I thought about it but got lost in the process. It's now added.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java (hunk 10 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java:53]: The intention is to always have the final query to be an instance of `ApiKeyBoolQueryBuilder` because it has special handling to disable certain fields on the SearchExecutionContext level. So yes this is necessary because it converts a normal boolQueryBuilder to a `ApiKeyBoolQueryBuilder`.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java (hunk 11 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java:77]: I went back and forth between `ids` and `term`. I understand the `ids` approach exposes the internal knowledge of an API Key document use the API Key ID as its document ID. And this is non-ideal. But the `term` approach has its own problem:
* How do we handle case-sensitivity? Is it not allowed? It is not obvious and could be trappy either way.
* It's not possible to support prefix, wildcard etc for the `_id` field, but it is again not obivious just from the Query DSL side.

(I remember I had some other reasons as well, but I cannot remember them right now.)

The bottom line is that the ID field will be an anonamly regardless of the approach. So we need to choose something that are less concerning. If "exposing the doc ID" is a big enough concern, I am happy to take the `term` (`terms`) approach.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ManageOwnApiKeyClusterPrivilege.java (hunk 1 above)
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ManageOwnApiKeyClusterPrivilege.java:82]: I assume your example is about the interceptor approach. We don't have such an interceptor for cluster actions, but I assume it would work similarly to the interceptors for index requests. The interface is

```java
void intercept(
  RequestInfo requestInfo, 
  AuthorizationEngine authorizationEngine, 
  AuthorizationInfo authorizationInfo,
  ActionListener<Void> listener);
```

Are you concerning about that, inside this method, we need to only rely on the fact that both `AuthorizationEngine` and `AuthorizationInfo` are interfaces and we need do it in a way that does not make `RBACEngine` and our `Role` special for at least in this method? It's challenging even with the "separate action name" approach. None of the existing methods fit this purpose really well. We can sorta fit it into `AuthorizationEngine#authorizeClusterAction`. But it would be better if we can add a new method to `AuthorizationEngine`. I think we'll need high bandwidth conversation on this.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java (hunk 10 above)
Comment (tvernum) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java:53]: But can't we just have the `if` case used all the time? Why doesn't that do what we want?
```
           QueryBuilder processedQuery = doProcess(queryBuilder);
           finalQuery.must(processedQuery);
```

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java (hunk 11 above)
Comment (tvernum) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java:77]: Thanks for the explanation.
I assume if in the future the id for the API Key document wasn't exactly the same as the API Key id, (but the key id was a field inside the doc) then we could translate an `ids` query into a term query.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java (hunk 10 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java:53]: Ah ok. Sorry I see what you mean now. You are right! The must clause can simply accept any query builder including a bool query (now it's kinda obvious after you pointed it out ...) It's a better approach and I assume the end results are the same after lucene level query rewrite. Will update. Thanks!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/apikey/QueryApiKeyResponse.java (hunk 3 above)
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/apikey/QueryApiKeyResponse.java]: Removed. It was a copy/paste from GetApiKeyResponse, which has its usage in RestGetApiKeyActionTests. But even for that I don't think it's necessary.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ManageOwnApiKeyClusterPrivilege.java (hunk 1 above)
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ManageOwnApiKeyClusterPrivilege.java]: I came up with a [new approach](https://github.com/elastic/elasticsearch/pull/75335/commits/786daad61b6a54859083feb3c1f9613911fbccef) inspired by your comment. The idea is to authorize the action a second time if it fails for the first time and it is the QueryApiKey action. I think this approach is promising because:
* It is simple. A few lines change in `AuthorizationService`
* In theory works with arbitrary `AuthorizationEngine` and `AuthorizationInfo`
* Does not need a separate action name and works kinda smoothly with the `manage_own_api_key` privilege.

It has the overhead of a second authorization for the QueryApiKey action. But I think it is acceptable because it authorizing on name is pretty fast and it does not affect any other actions.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java [hunk 14]
```diff
@@ -273,13 +275,25 @@ private void authorizeAction(final RequestInfo requestInfo, final String request
...
-            authzEngine.authorizeClusterAction(requestInfo, authzInfo, clusterAuthzListener);
+            authzEngine.authorizeClusterAction(requestInfo, authzInfo, ActionListener.wrap(result -> {
+                if (false == result.isGranted() && QueryApiKeyAction.NAME.equals(action)) {
+                    assert request instanceof QueryApiKeyRequest : "request does not match action";
+                    final QueryApiKeyRequest queryApiKeyRequest = (QueryApiKeyRequest) request;
+                    if (false == queryApiKeyRequest.isFilterForCurrentUser()) {
+                        queryApiKeyRequest.setFilterForCurrentUser();
+                        authzEngine.authorizeClusterAction(requestInfo, authzInfo, clusterAuthzListener);
+                        return;
+                    }
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java:293]: I like this option better.
It's still not perfect, so next time we do this, we should think about a more general solution, but this is good for the time being.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+
+public class ApiKeyBoolQueryBuilder extends BoolQueryBuilder {
+
+    // Field names allowed at the index level
+    private static final Set<String> ALLOWED_EXACT_INDEX_FIELD_NAMES =
+        Set.of("doc_type", "name", "api_key_invalidated", "creation_time", "expiration_time");
+
+    private ApiKeyBoolQueryBuilder() {}
+
+    public static ApiKeyBoolQueryBuilder build(QueryBuilder queryBuilder, Authentication authentication) {
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java:39]: Can we javadoc this method?

**Review by tvernum — APPROVED** [review]

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java (hunk 15 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java]: Good point. Added now.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java (hunk 11 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java:84]: Yes it's my thinking as well. Translate `ids` to a `term` query (in future) is easy, but not so much the other way around.

**Comment by ywangd** [comment]
@elasticmachine update branch
